Add PostAggregators to generator cache keys for top-n queries

jihoonson opened the pull request:
This PR is to fix #3719.

<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/druid-io/druid/3899)
<!-- Reviewable:end -->

drcrallen commented on processing/src/test/java/io/druid/query/cache/CacheKeyBuilderTest.java:
```diff
@@ -0,0 +1,121 @@
+import static org.junit.Assert.assertTrue;
+
+@RunWith(Parameterized.class)
+public class CacheKeyBuilderTest
```
Can you add a test for the following:

Item 1:
String1 : `a`
String2 : `b`

Item 2:
String 1:`ab`

jihoonson replied:
I added tests. Thanks.

nishantmonu51 commented on processing/src/main/java/io/druid/query/cache/CacheKeyBuilder.java:
```diff
@@ -0,0 +1,175 @@
+
+  public CacheKeyBuilder appendCacheableList(List<? extends Cacheable> inputs)
+  {
+    for (Cacheable input : inputs) {
```
can the list be null in any case? We might as well just add null a check here.

jihoonson replied:
I don't see any case now, but it will be helpful for future uses. I'll add it.

drcrallen commented on processing/src/main/java/io/druid/query/cache/CacheKeyBuilder.java:
```diff
@@ -0,0 +1,175 @@
+  public CacheKeyBuilder appendStringList(List<String> input)
+  {
+    for (String eachStr : input) {
+      appendItem(stringToByteArray(eachStr));
```
Ok, I'll be more explicit here.

If you call 

```java
appendStringList(ImmutalbeList.of("a","b"));
```

in one cache key builder, and 

```java
appendStringList(ImmutableList.of("ab"))
```

in another, they will both evaluate (incorrectly) to the same cache key, if I'm reading the code correctly. All strings should be terminated with a byte that is NOT a valid UTF-8 character start (like `0xFF`)

jihoonson replied:
@drcrallen, thanks. I'll add tests.

@gianm, thank you for the good suggestion. I'm concerned with the cache key size. How about adding some keys for each type like below?

```java
static final byte BYTE_KEY = 0;
static final byte BYTE_ARRAY_KEY = 1;
...
static final byte INT_KEY = 10;
static final byte INT_ARRAY_KEY = 11;
...

public CacheKeyBuilder appendByte(byte input)
{
  appendItem(BYTE_KEY, byteToByteArray(input));
  return this;
}

public CacheKeyBuilder appendByteArray(byte[] input)
{
  appendItem(BYTE_ARRAY_KEY, input);
  return this;
}

private void appendItem(byte key, byte[] bytes)
{
  ...
}
...
```

gianm replied:
@jihoonson, that sounds like a good way to save some space.

jihoonson replied:
I realized that both type keys and list sizes are needed to avoid this problem. I added them to CacheKeyBuilder.

fjy commented:
👍

gianm commented on server/src/test/java/io/druid/client/CachingClusteredClientTest.java:
```diff
@@ -205,7 +205,7 @@
   );
   private static final List<AggregatorFactory> RENAMED_AGGS = Arrays.asList(
-      new CountAggregatorFactory("rows2"),
+      new CountAggregatorFactory("rows"),
```
What's the idea behind this change, and the makeTopNResults -> makeTopNResultsWithoutRename change? It's not clear to me why the new test is better.

jihoonson replied:
This was not for a better test, but to make unit tests passed.

```CachingClusteredClientTest.testTopN*``` tests check the cached result, and here the same post aggregator should be used except their names. So, I just changed test queries to use ```POST_AGG``` which is also used for caching like [here](https://github.com/druid-io/druid/pull/3899/files/ddb4eee8859b1bd5d433b2df92e742e2a2ee9dc0#diff-7254bc82bb1d7b6afc78e03268c96806R821). 

To make ```POST_AGG``` work properly, the ```FieldAccessPostAggregator.fieldName``` should be the one specified in ```AGGS```. So, I changed the name of ```CountAggregatorFactory``` from ```rows2``` to ```rows```.

Since there is another renamed field ```impers2``` in ```RENAMED_AGG```, I thought it would not be a problem in other tests for group-by with renaming. If not, please let me know.

The change in makeTopNResultsWithoutRename() is just for my convenience. If its name is verbose, I'll change it.

gianm replied:
Ah I see what's going on. There's nothing wrong with verbose methods (usually I prefer them) so that's fine.

jihoonson commented on processing/src/main/java/io/druid/query/groupby/GroupByQueryQueryToolChest.java:
```diff
@@ -356,45 +355,13 @@ public String apply(DimensionSpec input)
+            .appendCacheable(query.getGranularity())
+            .appendCacheable(query.getDimFilter())
+            .appendCacheableList(query.getAggregatorSpecs())
+            .appendCacheableList(query.getDimensions())
```
Thanks. It's my bad. I'll fix it.

After #3873, the group by result can be different if the limit-push-down is enabled and the ordering keys contain non-dimensions. I think this should also be considered to make the cache key.

jon-wei replied:
The limit push down in #3873 is applied when merging the per-segment results, so taking the limit out of the cache key should be fine there as well

gianm commented on processing/src/main/java/io/druid/query/topn/TopNQueryQueryToolChest.java:
```diff
@@ -306,27 +303,15 @@ public String apply(@Nullable AggregatorFactory input)
+            .appendCacheable(query.getGranularity())
+            .appendCacheable(query.getDimensionsFilter())
+            .appendCacheableList(query.getAggregatorSpecs())
+            .appendCacheableList(query.getPostAggregatorSpecs())
```
Until now, just changing the post-aggregators on a topN (maybe adding new arithmetic, something like that) would still allow the cache to be used. This is a bug if we're sorting by the post-aggregator but is fine otherwise.

What do you think about changing the cache key generation so it only includes post-aggs we're sorting on (and any transitive dependencies based on `getDependentFields`)?

jihoonson replied:
Sounds good. I'll do.

jihoonson commented on processing/src/main/java/io/druid/query/aggregation/post/PostAggregatorIds.java:
```diff
@@ -0,0 +1,44 @@
+  public static final byte LONG_GREATEST = 11;
+  public static final byte LONG_LEAST = 12;
+  public static final byte MAX = 13;
+  public static final byte MIN = 14;
```
Changed.

jihoonson commented:
@drcrallen, @gianm thanks for your review. I addressed comments.
Additionally, I added the below feature to CacheKeyBuilder.
* When a collection is appended to CacheKeyBuilder, that collection is sorted by its items' byte representation. This is to guarantee that collections of same items but in different orders make the same result

gianm commented on processing/src/main/java/io/druid/query/topn/TopNQueryQueryToolChest.java:
```diff
@@ -291,6 +291,76 @@ public String apply(@Nullable AggregatorFactory input)
   }
 
+  @VisibleForTesting
+  static Collection<PostAggregator> findPostAggregatorsForSort(
```
This looks redundant to the already existing `private static List<PostAggregator> prunePostAggregators(TopNQuery query)` which calls `AggregatorUtil.pruneDependentPostAgg` (used for similar functionality in other parts of the topN system)

jihoonson replied:
Thanks. I changed to use that method.

gianm commented:
@jihoonson, thx for the update. had a couple further comments on the new stuff.

jihoonson commented on processing/src/main/java/io/druid/query/cache/CacheKeyBuilder.java:
```diff
@@ -0,0 +1,251 @@
+        byteArrayList.add(byteArray);
+      }
+
+      // Sort the byte array list to guarantee that collections of same items but in different orders make the same result
```
Good catch. Thanks. I added appendCacheablesIgnoringOrder(). 
I checked which one of appendCacheables and appendCacheablesIgnoringOrder should be called when building cache keys, but am not sure for SketchSetPostAggregator. Please verify it.

jihoonson commented:
@gianm, thanks. I addressed your comments. 

I think we need more tests to verify query caching with various types of fields, aggregators, post aggregators, etc. To make the tests concrete, I think it would be better to add more classes which are specialized for each query type. These tests are parameterized with original query, renamed parameters, and parameters of different order. However, this will cause a huge change in tests. I'm not sure about adding these tests in this PR. Maybe better to add in a following PR? What do you think?

gianm commented:
That test sounds great -- we don't have amazing coverage of caching and cache key right now. I think it makes sense to do it in a followup PR.

gianm commented on extensions-core/datasketches/src/main/java/io/druid/query/aggregation/datasketches/theta/SketchSetPostAggregator.java:
```diff
@@ -165,10 +165,29 @@ public int hashCode()
+    return builder.build();
+  }
+
+  private static boolean preserveFieldOrder(SketchHolder.Func func)
```
Suggest calling this `preserveFieldOrderInCacheKey` for better clarity.

gianm commented on processing/src/main/java/io/druid/query/aggregation/post/ArithmeticPostAggregator.java:
```diff
@@ -163,6 +169,21 @@ public String toString()
            '}';
   }
 
+  private static boolean preserveFieldOrder(Ops op)
```
Similar, suggest calling this `preserveFieldOrderInCacheKey`.

gianm commented on processing/src/main/java/io/druid/query/topn/TopNQueryQueryToolChest.java:
```diff
@@ -100,7 +97,7 @@ public String apply(@Nullable AggregatorFactory input)
   }
 
-  private static List<PostAggregator> prunePostAggregators(TopNQuery query)
+  public static List<PostAggregator> prunePostAggregators(TopNQuery query)
```
This doesn't need to be public, it's only called from within this file.

jihoonson commented:
@gianm, fixed. Thanks!

drcrallen commented on processing/src/main/java/io/druid/query/cache/CacheKeyBuilder.java:
```diff
@@ -0,0 +1,268 @@
+import java.util.Iterator;
+import java.util.List;
+
+public class CacheKeyBuilder
```
Can you add docs here for a high level overview of how this class works?

drcrallen commented on processing/src/main/java/io/druid/query/groupby/having/HavingSpec.java:
```diff
@@ -43,20 +44,20 @@
   // for testing.
-  public static final HavingSpec NEVER = new NeverHavingSpec();
-  public static final HavingSpec ALWAYS = new AlwaysHavingSpec();
+  HavingSpec NEVER = new NeverHavingSpec();
```
why this change?

gianm replied:
They don't do anything for interface fields.

drcrallen replied:
gotcha

drcrallen commented on processing/src/main/java/io/druid/query/topn/TopNQueryQueryToolChest.java:
```diff
@@ -306,27 +305,22 @@ public String apply(@Nullable AggregatorFactory input)
+            .appendCacheable(query.getDimensionsFilter())
+            .appendCacheablesIgnoringOrder(query.getAggregatorSpecs());
+
+        final List<PostAggregator> postAggregators = prunePostAggregators(query);
```
Does this mean that if I change post aggregators that are not part of the sorted that I will always get a cached result?

gianm replied:
That's the idea: only postaggs involved in the sorting could affect the topN by-segment results, so they're the only ones that need to be in the cache key.

drcrallen replied:
oh, duh, because the post-aggs can be calculated from the cached results.

Are they? My mind is drawing a blank there. Do post-aggs get calculated AFTER cached results?

gianm replied:
Yes, they are calculated after pulling from cache. But in this case, a postagg can affect the sort order of cached results, even if it's not actually _in_ the cached results. So it still needs to be part of the topN cache key (to fix #3719).

drcrallen replied:
Had to go look myself to make sure, `io.druid.client.CachingClusteredClient#mergeCachedAndUncachedSequences` and `io.druid.client.CachingQueryRunner#run` seem to be able to handle it but holy cow, I forgot how convoluted the data paths are for actually running a query.

drcrallen commented:
Requesting more class docs for the cache helper, but not strong enough to block the PR

gianm commented:
Going to merge this one then, but @jihoonson could you please do a follow up adding the javadocs that @drcrallen mentioned to CacheKeyBuilder? thx

jihoonson commented:
@gianm, @drcrallen thanks for your review. I opened #3933 for adding docs.
